### ISSUE TITLE
Persistence. Snapshots not working if there is no event sourcing

### ISSUE THREAD

**Issue by VioletTape** [AsynkronIT/protoactor-dotnet#839]
## Expected Behavior

Persistence class offers several ways to persist a state of an actor, one of which is to use snapshots only. In case of using this method `WithSnapshotting()` I'm expecting that snapshots taken and restored by request. 

I expect that with this strategy the system creates a single snapshot and updates the same object in persistence storage without creating a new version every time. 

## Actual Behavior

Unfortunately, if one attempts to create a second snapshot it won't work and fail with the message that such a key already exists (tested on MongoDB).

The reason is in logic inside [RecoverStateAsync()](https://github.com/AsynkronIT/protoactor-dotnet/blob/d4d2190dd131cd00a40be787856311603bf367bc/src/Proto.Persistence/Persistence.cs#L118) that assume about events presence to handle index for snapshots. 

Very likely the whole logic about persistence assumes that ES is used. 

The quick fix might be adding something like 
```csharp
if (_eventStore is NoEventStore) {
    Index++;
    return;
}
```

But it won't solve the problem that every time there would be a new snapshot entry in a storage.

## Steps to Reproduce the Problem

  1. Setup persistence with WithSnapshotting().
  2. Save object with PersistSnapshotAsync()
  3. Retrieve object RecoverStateAsync()
  4. Save object with PersistSnapshotAsync() - exception happens

Also all test cases for persistence use `WithEventSourcingAndSnapshotting()` strategy. If one creates an actor with `WithSnapshotting()` then the issue can be replayed. 

## Specifications

  - Version: v 0.16.1-alpha.0.8
  - Platform: Windows 
  - Subsystem: Persistence

**Comment by MarkEwer** [comment]
Perhaps a good solution to the problem of having multiple snapshots would be to just include an option to state how many snapshots you want to keep for a given actor.  That might look something like this:

```csharp
Persistence.WithSnapshotting(
  snapshotStore: _snapshotStoreProvider,
  actorId: "Actor-Id-42",
  applySnapshot: ApplySnapshot,
  maxSnapshots: 5);
```

The proposed new parameter called `maxSnapshots` is an integer that indicates how many snapshots you wish to keep for this actor.  In this case, when we get to the sixth snapshot for "Actor-Id-42" then the persistence system would automatically remove the oldest snapshot and keep only five in the data storage at any given time. In my problem domain, I would want to keep a few so I could easily roll back to previous versions of the actor but in @VioletTape's case, it could be set to 1 so that only one snapshot would be kept in the system.

**Comment by rogeralsing** [comment]
We have the new `IKeyValueStore<T>` which is aimed at this.
There is also a concrete impl, `ConcurrentKeyValueStore<T>` which limits internal calls using an `AsyncSemaphore`.

We can probably add KeyValueStore implementations to the persistence libraries to support this specific usecase as it is quite different from the idea of eventsourcing+snapshotting.

cc @alexeyzimarev thoughts? separate impls, or make persistence able to do single snapshot reads/writes w/o any events?

**Comment by rogeralsing** [comment]
After discussions with @alexeyzimarev.

One good way forward would be to allow to pass in strategies/configs to the Snapshot provider itself.
E.g. stating how many snapshots you want to keep or if you want to use TTL for snapshots etc.

**Comment by jstnlef** [comment]
Hey there. I was wondering how long it would take to possibly get a fix on this. I was the one who brought this up initially and am looking to get my first proto.actor based service out to our customers ASAP. I mostly ask in order to make the decision to either wait for this fix or to work around the bug. Thanks!

**Comment by rogeralsing** [comment]
PR closed the issue, reopened it again as there are more things to do still.

**At least the Index issue is fixed and available in version 0.18.1-alpha.0.30.**
